Content.Shared/Chat/V2/Moderation/SimpleCensor.cs

```diff
@@ -0,0 +1,272 @@
+/// </summary>
+public sealed class SimpleCensor
+{
+    private bool _sanitizeSpecialCharacters;
+    private bool _sanitizeL33TSpeak;
+    private bool _sanitizeSpaces;
+
+    private FrozenDictionary<char, char> _characterReplacements = new Dictionary<char, char>().ToFrozenDictionary();
+
+    private List<string> _censoredWords = new();
+    private List<string> _falsePositives = new();
+    private List<string> _falseNegatives = new();
+    private HashSet<UnicodeRange> _unicodeRanges = new();
+
+    public bool Censor(string input, out string output, char replaceWith = '*')
+    {
+        output = Censor(input, replaceWith);
+
+        return !string.IsNullOrEmpty(output);
+    }
+
+    public SimpleCensor WithSanitizeL33TSpeak()
+    {
+        _sanitizeL33TSpeak = true;
+
+        return BuildCharacterReplacements();
+    }
+
+    public SimpleCensor WithSanitizeSpecialCharacters()
+    {
+        _sanitizeSpecialCharacters = true;
+
+        return BuildCharacterReplacements();
+    }
+
+    public SimpleCensor WithRanges(HashSet<UnicodeRange> ranges)
+    {
+        _unicodeRanges = ranges;
+
+        return this;
+    }
+
+    public SimpleCensor WithSanitizeSpaces()
+    {
+        _sanitizeSpaces = true;
+
+        return this;
+    }
+
+    public SimpleCensor WithCustomDictionary(List<string> naughtyWords, List<string>? falsePositives, List<string>? falseNegatives)
+    {
+        _censoredWords = naughtyWords;
+
+        if (falsePositives != null)
+            _falsePositives = falsePositives;
+
+        if (falseNegatives != null)
+            _falseNegatives = falseNegatives;
+
+        return this;
+    }
+
+    public SimpleCensor WithCustomCharacterReplacements(Dictionary<char, char> replacements)
+    {
+        _characterReplacements = replacements.ToFrozenDictionary();
+
+        return this;
+    }
+
+    public string Censor(string input, char replaceWith = '*')
+    {
+        input = _unicodeRanges.Count > 0 ? ApplyUnicodeRangeAllowList(input) : input;
+
+        var censored = input.ToCharArray();
+
+        input = Sanitize(input, out var originalIndexes, true);
+        CheckProfanity(ref input, ref originalIndexes, ref censored, _falseNegatives, replaceWith);
+        RemoveFalsePositives(ref input, ref originalIndexes);
+        CheckProfanity(ref input, ref originalIndexes, ref censored, _censoredWords, replaceWith);
+
+        return new string(censored);
+    }
+
+    private void CheckProfanity(ref string input, ref List<int> originalIndexes, ref char[] censored, List<string> words, char replaceWith = '*')
+    {
+        foreach (var word in words)
+        {
+            var wordLength = word.Length;
+            var currentIndex = 0;
+
+            for (var foundIndex = input[currentIndex..].IndexOf(word, StringComparison.OrdinalIgnoreCase); foundIndex > -1; foundIndex = input[currentIndex..].IndexOf(word, StringComparison.OrdinalIgnoreCase))
```

Comment:
`input[currentIndex..]` allocates a string copy, you should use the `IndexOf()` variant that allows you to select a start position.